components/EditPublicMessagePopup.js

```diff
@@ -0,0 +1,304 @@
+        filter: drop-shadow(-4px 0px 4px rgba(20, 20, 20, 0.09));
+        transform: translateY(-50%);
+      }
+    `}
+`;
+
+const EditPublicMessageMutation = gql`
+  mutation EditPublicMessageMutation($FromCollectiveId: Int!, $CollectiveId: Int!, $message: String) {
+    editPublicMessage(FromCollectiveId: $FromCollectiveId, CollectiveId: $CollectiveId, message: $message) {
+      id
+      publicMessage
+      tier {
+        id
+      }
+      collective {
+        slug
+      }
+    }
+  }
+`;
+
+const messages = defineMessages({
+  publicMessagePlaceholder: {
+    id: 'contribute.publicMessage.placeholder',
+    defaultMessage: 'Motivate others to contribute in 140 characters :) ...',
+  },
+});
+
+const PUBLIC_MESSAGE_MAX_LENGTH = 140;
+
+function EditPublicMessagePopup({ fromCollectiveId, collectiveId, cardRef, onClose, message, intl }) {
+  // Root element outside of React where the popup will be rendered.
+  const rootElementRef = useRef();
+  // Popup root element reference
+  const popupRef = useRef();
+  const [editMessagePopupPosition, setEditMessagePopupPosition] = useState({
+    x: 0,
+    y: 0,
+    left: false,
+    right: false,
+  });
+  const [messageDraft, setMessageDraft] = useState(message || '');
+
+  /** Add the root element to the document */
+  useLayoutEffect(() => {
+    document.body.appendChild(rootElementRef.current);
+    return () => rootElementRef.current.remove();
+  }, []);
+
+  /** Determine the position to render the popup relative to the contributor card */
+  useLayoutEffect(() => {
+    const cardRect = cardRef.current.getBoundingClientRect();
+    const popupRect = popupRef.current.getBoundingClientRect();
+    /** Check if popup fits to the right */
+    const viewportWithWithoutScrollbar = document.body.clientWidth;
+    const cardX = window.scrollX + cardRect.x;
+    const cardY = window.scrollY + cardRect.y;
+    if (cardRect.x + cardRect.width + popupRect.width + POPUP_ARROW_WIDTH <= viewportWithWithoutScrollbar) {
+      const x = cardX + cardRect.width + POPUP_ARROW_WIDTH;
+      setEditMessagePopupPosition({ x, y: cardY, right: true });
+      // /** Check if popup fits to the left */
+    } else if (cardRect.x - popupRect.width - POPUP_ARROW_WIDTH >= 0) {
+      const x = cardX - popupRect.width - POPUP_ARROW_WIDTH;
+      setEditMessagePopupPosition({ x, y: cardY, left: true });
+      // /** Position popup at the bottom */
+    } else {
+      const popupHalfWidth = popupRect.width / 2;
+      const cardHalfWidth = cardRect.width / 2;
+      let x = cardX - popupHalfWidth + cardHalfWidth;
+      /** Add vertical offset */
+      const y = cardY + cardRect.height + 10;
+      /** Check if popup will fit horizontally. Adjust if it doesn't */
+      if (x < 0) {
+        x = 0;
+      } else if (x + popupRect.width > viewportWithWithoutScrollbar) {
+        x -= x + popupRect.width - viewportWithWithoutScrollbar;
+      }
+      setEditMessagePopupPosition({ x, y });
+    }
+  }, []);
```

Comment:
Using `StyledTooltip` doesn't solve one of the problems that pushed me to implement my own solution.

As you can see in the first image that scrollbar is not the one from the browser but from the grid component. The problem is that the modal is being rendered inside the grid along side the other components as you can see in the second picture. To solve it I used [portals](https://reactjs.org/docs/portals.html). I am rendering the popup in a `div` that is a direct child of `body` and positioning the element relative to the whole viewport.

![Screen Shot 2019-11-05 at 4 04 16 PM](https://user-images.githubusercontent.com/1315054/68244087-6f861880-ffea-11e9-8c93-25b5f959cf93.png)

![Screen Shot 2019-11-05 at 4 04 58 PM](https://user-images.githubusercontent.com/1315054/68244160-99d7d600-ffea-11e9-97e2-1c96fa195436.png)

> This code doesn't seem to work for mobile (see below). 
The component doesn't have enough space to be rendered to the left or right of the card so, it is being rendered bellow. This is the same behavior that the modal already had in the order success page. The only new behavior here is the ability to render the modal to the left if there is enough space. This was not a problem in the order success page because the card was centered vertically.

Reply:
Also I see that [react-tooltip@3.11.1 minified and gziped is 7.1 kB](https://bundlephobia.com/result?p=react-tooltip@3.11.1), not that much, but to implement something similar took me like 30 lines of code. Maybe it would be a good chance to evaluate if a custom solution/in house would be good enough and save some bytes.

Reply:
Alright, thanks for all the details. That makes sense, and good job for taking the extra-step to better position the popup.

 > Also I see that react-tooltip@3.11.1 minified and gziped is 7.1 kB, not that much, but to implement something similar took me like 30 lines of code. Maybe it would be a good chance to evaluate if a custom solution/in house would be good enough and save some bytes.

That's right. `StyledTooltip` was a first step towards abstracting tooltips under a unique API. An in house solution would indeed make sense and that'd be an opportunity to implement a more responsive behavior.

Reply:
If you want we can create an issue to discuss about the api for a tooltip in house solution. I am thinking about a `Positioner` component. All the logic for positioning can be in it, and all the tooltips and modals of the app can use it.